loader/db.go

```diff
@@ -119,48 +122,54 @@ func (conn *Conn) executeSQLCustomRetry(sqls []string, enableRetry bool, isRetry
 }
 
 func executeSQLImp(db *sql.DB, sqls []string) error {
-	var (
-		err error
-		txn *sql.Tx
-		res sql.Result
-	)
-
-	txn, err = db.Begin()
-	if err != nil {
-		log.Errorf("exec sqls[%-.100v] begin failed %v", sqls, errors.ErrorStack(err))
-		return err
-	}
-
-	for i := range sqls {
-		log.Debugf("[exec][sql]%-.200v", sqls[i])
-		res, err = txn.Exec(sqls[i])
+	for {
+		var (
+			err error
+			txn *sql.Tx
+			res sql.Result
+		)
+
+		txn, err = db.Begin()
 		if err != nil {
-			log.Warnf("[exec][sql]%-.100v[error]%v", sqls[i], err)
-			rerr := txn.Rollback()
-			if rerr != nil {
-				log.Errorf("[exec][sql]%-.100s[error]%v", sqls, rerr)
-			}
+			log.Errorf("exec sqls[%-.100v] begin failed %v", sqls, errors.ErrorStack(err))
 			return err
 		}
-		// check update checkpoint successful or not
-		if i == 2 {
-			row, err1 := res.RowsAffected()
-			if err1 != nil {
-				log.Warnf("exec sql %s get rows affected error %s", sqls[i], err1)
-				continue
+
+		for i := range sqls {
+			log.Debugf("[exec][sql]%-.200v", sqls[i])
+			res, err = txn.Exec(sqls[i])
+			if err != nil {
+				log.Warnf("[exec][sql]%-.100v[error]%v", sqls[i], err)
+				rerr := txn.Rollback()
+				if rerr != nil {
+					log.Errorf("[exec][sql]%-.100s[error]%v", sqls, rerr)
+				}
+				return err
 			}
-			if row != 1 {
-				log.Warnf("update checkpoint affected rows %d", row)
+			// check update checkpoint successful or not
+			if i == 2 {
+				row, err1 := res.RowsAffected()
+				if err1 != nil {
+					log.Warnf("exec sql %s get rows affected error %s", sqls[i], err1)
+					continue
+				}
+				if row != 1 {
+					log.Warnf("update checkpoint affected rows %d", row)
+				}
 			}
 		}
-	}
 
-	err = txn.Commit()
-	if err != nil {
-		return errors.Trace(err)
+		err = txn.Commit()
+		if err != nil {
+			if !strings.Contains(err.Error(), txnRetryableMark) {
```

Comment:
how about add a function to judge whether this err is retryable?

Reply:
Still loading data. So is there other place that should test this and retry indefinitely?

Reply:
Found there is similar code in syncer. Let me move it to pkg/utils/db.go so it can be reused.

Reply:
Done, and I found there is error code checking implementation in pkg/utils/db.go. So it's testing code instead of tricky pattern matching.